connectors/slack/template/index.js

```diff
@@ -1,24 +1,321 @@
+                //         });
+                //     });
+            } else {
+                return cb(null, { status: 404 });
+            }
+        } else {
+            return cb(null, {
+                status: 404,
+                body: 'Connector is not configured for Slack Events API',
+            });
+        }
+    }
+};
 
+function handleEventIngest(ctx, cb) {
+    debugTime('handleEventIngest');
+    debugTimeLog(
+        'handleEventIngest',
+        `starting handleEventIngest ${ctx.query.dispatch !== undefined ? '(dispatch) ' : '(ingest) '}for`,
+        ctx.body && ctx.body.event && ctx.body.event.text
+    );
 
-Completion - 302 redirect to returnTo URL with query params
-- [required] status (success or error)
-- [required] data (augmented settings from initialization, or new settings if initialization did not provide any)
-- [required if present on initialization] state (from initialization)
+    try {
+        verifyRequestSignature({
+            signingSecret: ctx.configuration.slack_signing_secret,
+            requestSignature: ctx.headers['x-slack-signature'],
+            requestTimestamp: parseInt(ctx.headers['x-slack-request-timestamp'], 10),
+            body: JSON.stringify(ctx.body).replace(/\//g, '\\/'),
+        });
+    } catch (e) {
+        Sdk.debug('New event signature verification failed');
+        Sdk.debug(JSON.stringify(ctx.body));
+        cb(null, { status: 403 });
+        return;
+    }
 
-Whatever happens between initialization and completion is an implementation detail, but in a general case may involve a number of redirects.
+    Sdk.debug('New event signature verified');
+    debugTimeLog('handleEventIngest', 'signature verified');
 
-/**
-* @param ctx {FusebitContext}
-*/
-module.exports = Sdk.createSettingsManager(configure);
+    return ctx.query.dispatch !== undefined ? dispatch(ctx, cb) : respondToSlack(ctx, cb);
+}
+
+function respondToSlack(ctx, cb) {
+    debugTime('respondToSlack');
+    debugTimeLog('respondToSlack', 'starting respondToSlack');
+    let slackTimestamp = +ctx.headers['x-slack-request-timestamp'] * 1000;
+    Sdk.debug('Slack request time', slackTimestamp);
+
+    // URL Verification Challenge
+    if (ctx.body && ctx.body.type && ctx.body.type === 'url_verification') {
+        Sdk.debug('URL Verification Challenge');
+        cb(null, { body: { challenge: ctx.body.challenge } });
+        return;
+    }
+
+    // Call back to itself to dispatch, wait 100ms, and confirm success to slack
+    // The async dispatch can run beyond the Slack's 3s limit
+
+    Superagent.post(ctx.baseUrl + '/events/ingest?dispatch')
+        .set(ctx.headers)
+        .send(ctx.body)
+        .then(function (result) {
+            debugTimeLog('respondToSlack', 'dispatch request to self completed with success');
+        })
+        .catch(function (error) {
+            debugTimeLog('respondToSlack', 'dispatch request to self completed with erroer', error);
```

Comment:
An error here should also result in an error back to slack, since obviously things are pretty broken and we want to yell as loudly as possible.

Reply:
spelling on error

Reply:
Yavor will add some comments to the code here indicating why the errors are eaten, and what would happen if they were returned, and how to debug problems in this space.

Reply:
Won't fix, we want to return 200 to Slack so it's easier to debug this